[vcpkg-tool-meson] fix #41869

vipcxj opened the pull request:
<!-- If your PR fixes issues, please note that here by adding "Fixes #NNNNNN." for each fixed issue on separate lines. -->

<!-- If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/. -->

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] SHA512s are updated for each updated download.
- [ ] The "supports" clause reflects platforms that may be fixed by this new version.
- [ ] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [ ] Any patches that are no longer applied are deleted from the port's directory.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

<!-- If this PR adds a new port, please uncomment and fill out this checklist:

- [ ] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [ ] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [ ] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html).
- [ ] The versioning scheme in `vcpkg.json` matches what upstream says.
- [ ] The license declaration in `vcpkg.json` matches what upstream says.
- [ ] The installed as the "copyright" file matches what upstream says.
- [ ] The source code of the component installed comes from an authoritative source.
- [ ] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [ ] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [ ] Only one version is in the new port's versions file.
- [ ] Only one version is added to each modified port's versions file.

END OF NEW PORT CHECKLIST (delete this line) -->

Only fixes ports built with meson, not other ports that use python directly. adding a python._pyd file will also fix this issue. However, python._pyd will completely overwrite sys.path and ignore all relevant environment variables, which may cause new bugs

Neumann-A commented on ports/vcpkg-tool-meson/vcpkg_install_meson.cmake:
```diff
@@ -30,6 +30,9 @@ function(vcpkg_install_meson)
                 vcpkg_add_to_path(PREPEND "${CURRENT_INSTALLED_DIR}/bin")
             endif()
         endif()
+        vcpkg_find_acquire_program(PYTHON3)
+        get_filename_component(PYTHON3_DIR "${PYTHON3}" DIRECTORY)
+        set(ENV{PYTHONHOME} ${PYTHON3_DIR})
```
vcpkg_install_meson just calls ninja. If the build requires python that is the responsibility of the portfile

Neumann-A replied:
> In my opinion

You didn't present facts you just presented an opinion.

Practically, meson can be a self-sustained binary without a proper python environment or even be called with a different python environment than is consumed by the project using meson.

As such if python is a requirement by the BUILD it has to be a DEPENDENCY in the manifest and thus using vcpkg_find_acquire_program would be wrong. 
The only exception to that is if it is just executing some python script as part of the build but even then the call needs to be in the portfile not here.

vipcxj replied:
"ports that use meson basically don't call python directly anymore" is my opinion, but "many ports that use meson basically don't call python directly anymore" is the fact.
If you are right, then why call vcpkg_find_acquire_program(PYTHON3) in vcpkg_configure_meson?
Assuming that vcpkg_configure_meson does need to call vcpkg_find_acquire_program(PYTHON3), it is not a fault for vcpkg_install_meson, which is often used with vcpkg_configure_meson, to call vcpkg_find_ acquire_program(PYTHON3) as well.

dg0yt replied:
> it is not a fault for vcpkg_install_meson, which is often used with vcpkg_configure_meson, to call vcpkg_find_ acquire_program(PYTHON3) as well.

It *is* wrong because it is *is not* the responsibility of `vcpkg_install_meson`.

vipcxj replied:
@dg0yt Why does vcpkg_configure_meson then execute vcpkg_find_ acquire_program(PYTHON3) but vcpkg_install_meson does not? I'll put aside the question of responsibility for now, from existing practice, vcpkg_configure_meson is always followed by vcpkg_install_meson. vcpkg_find_ acquire_program(PYTHON3) has already been called once, is there any side effect of calling it again? The python3 variables acquired by the two calls are inconsistent? Will the python be download again?

dg0yt replied:
> Why does vcpkg_configure_meson then execute vcpkg_find_ acquire_program(PYTHON3) but vcpkg_install_meson does not?

`vcpkg_configure_meson` runs meson which is a python program.  
`vcpkg_install_meson` runs ninja which a native program.

vipcxj replied:
@dg0yt So what are the specific side effects of calling vcpkg_find_acquire_program(PYTHON3) at vcpkg_install_meson? Because it's already called once in vcpkg_configure_meson, I can't understand why it can't be called again in vcpkg_install_meson. Are there examples of just using vcpkg_configure_meson without using vcpkg_install_meson? I would like you to refute me on a practical level. My goal in modifying vcpkg_install_meson is to make sure that the python that might be installed at vcpkg_configure_meson is used correctly. If the port is using a different python than vcpkg_configure_meson at the vcpkg_install_meson stage, that's an unusual piece of business, and I think that's the case where the port needs to be tweaked separately. To minimize the hassle, I think it would be a good idea to add a parameter switch to vcpkg_install_meson to control whether or not to introduce this environment variable, and default behavior is introduce it

dg0yt replied:
We don't enter that discussion. ninja doesn't need python, and so `vcpkg_install_meson` doesn't need it.

If the commands run by ninja need python, then the port has to deal with setting up python for that purpose. Even before running `vcpkg_configure_meson`.

vipcxj replied:
@dg0yt I really can't understand your logic. Please tell me what is the harm in my approach? Instead of just saying there is no point in discussing it. You guys always emphasize that it's not needed, but it's exactly what would be useful and solve the problem. If it needs to be forbidden, the only reason I can accept is that it may have some sort of side effect.

dg0yt replied:
You reject the reason but it doesn't change the fact. 
"Useful and solve the problem" is not enough to make it the right solution to a problem.
The undesired "side effect" is that it facilitates to hide the fact that a package depends on running python.

vipcxj replied:
“The undesired "side effect" is that it facilitates to hide the fact that a package depends on running python”
it's not the fact. 
The only "side effect" is introduce or override the env PYTHONHOME, it may infer the behavior of python. if no python is needed, it does not change anything.

vipcxj replied:
@dg0yt Let us to talk about facts, not dogma.
According to my experiments, after executing vcpkg_configure_meson on windows, when executing vcpkg_install_meson, the python installed by vcpkg is used, even though I already have a lot of other pythons installed on my computer.I think this is enough proof that vcpkg_configure_ meson has changed the behavior of vcpkg_install_meson at least on windows, so the side effects I mentioned above were not introduced by my pr. Currently I'm finding that on other platforms, instead of using python with vcpkg, python under /usr/bin is used, which is causing getting wrong PYTHONHOME. The solution I've come up with so far is to limit the above modification to windows, after all I've only encountered this bug on windows. i've installed a lot of python on linux as well, but this bug doesn't occur, this is probably due to the fact that linux avoids this problem by using the python that comes with the operating system directly

dg0yt replied:
Let's put everything in ports.cmake. Everything readily available for everybody. Useful and solves the problem.

vipcxj replied:
@dg0yt We are here to solve problems, so I hope we can discuss the reality. The reality is that almost all ports that use meson have this issue. Your idea is to modify all ports that use meson that might have problems, but this problem used to not exist. I don't know which modification introduced the problem, but at least it proves that those ports were fine in the first place. I really don't see the practicality of changing so many ports that work when you can fix the problem with a few lines of code.

dg0yt replied:
> The reality is that almost all ports that use meson have this issue.

Some ports don't have problems? So there is really no functional dependency.

 > Your idea is to modify all ports that use meson that might have problems

My "idea" is that ports which need python during the build step need to get it explicitly or via transitive dependencies. Just the same as with CMake or autotools.

 > this problem used to not exist

And where did it come from? From meson updates? From python update (https://github.com/microsoft/vcpkg/pull/41232)? Does one (meson) port introduce build-time python dependencies for some other (meson) ports?

 > I really don't see the practicality of changing so many ports that work when you can fix the problem with a few lines of code.

A few line of codes might be enough. This discussion is about `vcpkg_install_meson` needing modification. In particular when it would just assign the same value that was already assigned by `vcpkg_configure_meson`.

Does `vcpkg_find_acquire_program(PYTHON3)` do the right thing, reliably?  (#41874?)

Neumann-A replied:
Vcpkg python probably needs to pretend to be a venv python. Then it can install packages and be isolated.

vipcxj replied:
> Some ports don't have problems? So there is really no functional dependency.

I don't understand what you mean.
> My "idea" is that ports which need python during the build step need to get it explicitly or via transitive dependencies. Just the same as with CMake or autotools.

Builds of meson projects almost always use python, and since they use a certain python in the configure phase, it's only natural to use the same python if they continue to use it in the install phase. If they purposely use a different python, this is a special case and is generally handled specially. My pr has taken this special case into account by giving the port writer a choice. Also I want to emphasize again that my pr does not introduce a new python in the meson install phase, it introduces environment variables, which will be explained in more detail below
> And where did it come from? From meson updates? From python update (https://github.com/microsoft/vcpkg/pull/41232)? Does one (meson) port introduce build-time python dependencies for some other (meson) ports?

Same as above, I just changed an environment variable, and that environment variable should only take effect during vcpkg_confgure_meson and vcpkg_install_meson
> Does vcpkg_find_acquire_program(PYTHON3) do the right thing, reliably? (https://github.com/microsoft/vcpkg/pull/41874?)

This test just proved me right. My suspicion that using python._pyd might affect existing ports was confirmed by this experiment. And I think python._pyd has a worse problem, python._pyd completely prevents the port writer from doing anything to influence sys.path.

First of all, a clear definition of "no side-effects" should be provided to avoid several possible meanings
For me vcpkg_find_acquire_program(PYTHON3) has no side effects means that the
```cmake
vcpkg_find_acquire_program(PYTHON3)
```
equals
```cmake
vcpkg_find_acquire_program(PYTHON3)
vcpkg_find_acquire_program(PYTHON3)
```
vcpkg_find_acquire_program(PYTHON3) should, and must, have no side effects. Because during cmake configure, it can be repeatly invoked by multiple ports in an unknown order, and if there are side effects, it means bigger bugs, which are not included in this pr discussion

And then my second assumption: vcpkg_configure_meson is often used in conjunction with vcpkg_install_meson, maybe you can find a counterexample, but it's a rare case.
So one can exit with the following conclusion: since vcpkg_configure_meson has already called vcpkg_find_acquire_program(PYTHON3), calling vcpkg_find_acquire_program again in vcpkg_install_meson( PYTHON3) has no side effects, so my pr doesn't mess with any additional python dependencies in vcpkg_install_meson. The effect of my pr's changes in vcpkg_install_meson was limited to modifying an environment variable. If the port does use python in the vcpkg_install_meson phase, my environment variable affects the sys.path inference.
In the vast majority of cases there is no need to use more than one python at all, and it's a natural inference that the same python is used, so this environment variable correctly fixes the sys.path error. At least the result of ci proves this. If there are very special cases where this fix is problematic, I have also provided a parameter to cancel the introduction of this environment variable

Based on the current implementation, on windows, the python downloaded by vcpkg will always be used, and my pr gets its PYTHONHOME correctly. if this implementation will be modified in the future, that's not what this pr needs to focus on. I can only write the pr based on the current facts. on other platforms, vcpkg will choose the python provided by the system, which may be the one that comes with the OS or the one that comes with conda, and for the latter I can get the correct PYTHONHOME, but not for the former. The good thing is that on non-windows platforms, python's sys.path doesn't look like it needs to be fixed.

The last thing is about simulating a virtual environment to solve this problem. This seems like a perfect solution, but it just seems like it, in practice I imagine it would run into more issues, vcpkg isn't a python build tool, is it really necessary to put that much effort into it?

Neumann-A commented on scripts/cmake/vcpkg_install_meson.cmake:
```diff
@@ -30,6 +30,9 @@ function(vcpkg_install_meson)
                 vcpkg_add_to_path(PREPEND "${CURRENT_INSTALLED_DIR}/bin")
             endif()
         endif()
+        vcpkg_find_acquire_program(PYTHON3)
+        get_filename_component(PYTHON3_DIR "${PYTHON3}" DIRECTORY)
+        set(ENV{PYTHONHOME} ${PYTHON3_DIR})
```
same as above. also this is deprecated

Neumann-A commented on ports/vcpkg-tool-meson/vcpkg_configure_meson.cmake:
```diff
@@ -433,6 +433,7 @@ function(vcpkg_configure_meson)
 
     vcpkg_find_acquire_program(PYTHON3)
     get_filename_component(PYTHON3_DIR "${PYTHON3}" DIRECTORY)
+    set(ENV{PYTHONHOME} ${PYTHON3_DIR})
```
The meson call should be with `-I` which means this variables gets ignored?

vipcxj replied:
Environment variables are not ignored, at least when I tested it on windows. Without configuring environment variables, my local gstreamer fails to configure, with them added it succeeds.

Neumann-A commented:
> I don't remember which ports fail if this is commented.

The problem is you cannot properly install additional python packages.

Neumann-A commented:
Discussion in https://github.com/microsoft/vcpkg/pull/41871#discussion_r1823212205 is not resolved. Mentioning this here so that maintainers see it.

vipcxj commented:
> Discussion in [#41871 (comment)](https://github.com/microsoft/vcpkg/pull/41871#discussion_r1823212205) is not resolved. Mentioning this here so that maintainers see it.

Have resolved. vcpkg_find_acquire_program(PYTHON3) not exist in vcpkg_install_meson now. because python could be send as additional_binary to vcpkg_configure_meson which could not be achieved by vcpkg_find_acquire_program(PYTHON3) in vcpkg_install_meson

dg0yt commented on ports/vcpkg-tool-meson/vcpkg_install_meson.cmake:
```diff
@@ -30,6 +35,10 @@ function(vcpkg_install_meson)
         endif()
+        if(NOT arg_EXPORTED_PYTHON_HOME STREQUAL "" AND DEFINED ${arg_EXPORTED_PYTHON_HOME})
+            # arg_EXPORTED_PYTHON_HOME is defined by vcpkg_configure_meson to hold the python home.
+            set(ENV{PYTHONHOME} "${${arg_EXPORTED_PYTHON_HOME}}")
```
This is expected to have happened already before: In `vcpkg_configure_meson`.
So the whole modification to `vcpkg_install_meson` is redundant.

vipcxj replied:
I wouldn't have added it if it was redundant, without this code some ports that use python in vcpkg_install_meson will have sys.path problems, and this pr is there to fix that.
In my local testing, the environment variable is not global and the env defined in vcpkg_configure_meson does not affect vcpkg_install_meson. but I do wonder why vcpkg_add_to_path in vcpkg_configure_meson can affect vcpkg_ install_meson.

vipcxj replied:
If you want to test it locally on windows, install a version of python that doesn't match the version that comes with vcpkg and reproduce #41869 first. Most ports that use meson reproduce the problem, such as glib, gstreamer, etc. ports like xcb, which don't use meson, but instead use python directly, have the same problem. But this pr doesn't intend to solve this kind of problem.

vipcxj commented:
for port gobject-introspection, it always use python in "D:\vcpkg\downloads\tools\python\python-3.12.7-x64\python.exe" as meson main binary. (loged in meson.log) Even I delete all vcpkg_find_acquire_program(PYTHON3) in port vcpkg-tools-meson.
however, the port gobject-introspection send "python='D:/projects/test-vcpkg/build/vcpkg_installed/x64-windows/tools/python3/python.exe'" as additional binary. This cause port failed.
Could anyone tell me how meson decide the main python binary.

MonicaLiu0311 commented:
> for port gobject-introspection, it always use python in "D:\vcpkg\downloads\tools\python\python-3.12.7-x64\python.exe" as meson main binary. (loged in meson.log) Even I delete all vcpkg_find_acquire_program(PYTHON3) in port vcpkg-tools-meson. however, the port gobject-introspection send "python='D:/projects/test-vcpkg/build/vcpkg_installed/x64-windows/tools/python3/python.exe'" as additional binary. This cause port failed. Could anyone tell me how meson decide the main python binary.

Maybe [gobject-introspection/portfile.cmake#L47](https://github.com/microsoft/vcpkg/blob/master/ports/gobject-introspection/portfile.cmake#L47).

vipcxj commented:
> > for port gobject-introspection, it always use python in "D:\vcpkg\downloads\tools\python\python-3.12.7-x64\python.exe" as meson main binary. (loged in meson.log) Even I delete all vcpkg_find_acquire_program(PYTHON3) in port vcpkg-tools-meson. however, the port gobject-introspection send "python='D:/projects/test-vcpkg/build/vcpkg_installed/x64-windows/tools/python3/python.exe'" as additional binary. This cause port failed. Could anyone tell me how meson decide the main python binary.
>
> Maybe [gobject-introspection/portfile.cmake#L47](https://github.com/microsoft/vcpkg/blob/master/ports/gobject-introspection/portfile.cmake#L47).

No, this line is the additional binary mentioned in my question, but here python='D:/projects/test-vcpkg/build/vcpkg_installed/x64-windows/tools/python3/python.exe ', but the python I see based on the logs is "D:\vcpkg\downloads\tools\python\python-3.12.7-x64\python.exe"

dg0yt commented:
> the port gobject-introspection send "python='D:/projects/test-vcpkg/build/vcpkg_installed/x64-windows/tools/python3/python.exe'" as additional binary.

Ah, `gobject-introspection`, the port still waiting for proper cross-build support.
Why does it have to use the host python from the vcpkg port as an additional binary? A poor attempt to fix the original problem?

vipcxj commented:
> > the port gobject-introspection send "python='D:/projects/test-vcpkg/build/vcpkg_installed/x64-windows/tools/python3/python.exe'" as additional binary.
>
> Ah, `gobject-introspection`, the port still waiting for proper cross-build support. Why does it have to use the host python from the vcpkg port as an additional binary? A poor attempt to fix the original problem?

I just want this pr does not break other port. `gobject-introspection` is just a example. It exposes a problem that I'm having trouble understanding, which is how meson decide to use which python. In my original understanding, if a python is specified in the additional binary, then meson should use that python, and if not, find a python from the PATH, but this example disproves the above assumption. This may cause my logic for this pr to no longer be accurate.

vipcxj commented:
I have a lot of python installed on my host, both python312 installed directly, and the installation of memba which comes with python. so without vcpkg intervention, I don't think meson should pick python under downloads by itself. 

I tried to exclude the `vcpkg_find_acquire_program( PYTHON3)` interference, tried commenting out [ports/vcpkg-tool-meson/vcpkg-port-config.cmake#L47](https://github.com/microsoft/vcpkg/blob/2a3138723698306f4261632c92dc782f586167e3/ports/vcpkg-tool-meson/vcpkg-port-config.cmake#L47), and since the code here is just checking to see if the python version is greater than 3.7, I think it's safe to temporarily comment it out.

After that I fixed the previously mentioned bug and got `z_vcpkg_extract_python` to work. Since the specified python can be detected from the additional binaries used by `gobject-introspection`, `vcpkg_find_acquire_program(PYTHON3)` in `vcpkg_configure_meson` is also not called, and `PYTHON3` is specified to "${CURRENT_HOST_INSTALLED_DIR}/tools/python3/python${VCPKG_HOST_EXECUTABLE_SUFFIX}", `PYTHON3_ DIR` to "${CURRENT_HOST_INSTALLED_DIR}/tools/python3" and add it to the `PATH`

The appeal path is judged to be compliant, set the environment variable `PYTHONHOME`: "${CURRENT_HOST_INSTALLED_DIR}/tools/python3", export the cmake variable `VCPKG_PYTHON3_DIR`

The presence of `VCPKG_PYTHON3_DIR` is detected in vcpkg_install_meson, so set the environment variable `PYTHONHOME`

During the above process `vcpkg_find_acquire_program(PYTHON3)` was not called any more. But I found "Main binary: ${DOWNLOADS}/tools/python/pythonXXX/python.exe" in meson.log. Of course, since the PYTHONHOME is set to "${CURRENT_HOST_INSTALLED_DIR}/tools/python3", the build failed.

That's what happened to me last night.

In fact something happened before that struck me as odd. In a previous version of pr, `vcpkg_add_to_path(PREPEND "${PYTHON3_DIR}")` was placed in the `if scope`, which caused it to potentially not be executed. So ci did fail. 

When I looked at the logs, I found that in vcpkg_install_meson, meson was complaining that it couldn't find python. 

vcpkg_install_meson in this version of pr call neither `vcpkg_find_acquire_program(PYTHON3)` nor  `vcpkg_add_to_path(PREPEND "${PYTHON3_DIR}")`, so rightfully I assumed that it was the `vcpkg_add_to_path(PREPEND "${PYTHON3_DIR}")` in vcpkg_configure_meson that did the trick. 

When I moved `vcpkg_add_to_path(PREPEND "${PYTHON3_DIR}")` out of the `if scope`, the error disappeared, which confirmed my suspicions. But I don't think that `vcpkg_add_to_path(PREPEND "${PYTHON3_DIR}")`, located in vcpkg_configure_meson, should work on another scope that is completely unrelated!

Tonight i will try again, do more research.

dg0yt commented:
> I found that in `vcpkg_install_meson`, meson was complaining that it couldn't find python.

meson running again in the build/install step (i.e. appearing in `package-...log`)? Maybe via an build/install rule...

dg0yt commented:
> > `vcpkg_configure_meson` runs `MESON` which is actually `"${PYTHON3}" meson.py`.

> [scripts/cmake/vcpkg_find_acquire_program(MESON).cmake#L3](https://github.com/microsoft/vcpkg/blob/9d163f0612f75a49133cfd051f0ed997a24602d5/scripts/cmake/vcpkg_find_acquire_program(MESON).cmake#L3) `vcpkg_find_acquire_program(MESON)` uses the cmake variable `PYTHON3` as the interpreter for meson. So the order of `vcpkg_find_acquire_program(MESON)` and `vcpkg_find_acquire_program(PYTHON3)` is very important. To be more precise, `PYTHON3` must be determined before calling `vcpkg_find_acquire_program(MESON)`.

It is not meant to be necessary to set `PYTHON3` before calling `vcpkg_find_acquire_program(MESON)`. [scripts/cmake/vcpkg_find_acquire_program(MESON).cmake#L3](https://github.com/microsoft/vcpkg/blob/9d163f0612f75a49133cfd051f0ed997a24602d5/scripts/cmake/vcpkg_find_acquire_program(MESON).cmake#L3) sets the `interpreter` to `PYTHON3`, with the consequence that `vcpkg_find_acquire_program()` must also call `vcpkg_find_acquire_program(PYTHON3)`. That's how this variables is initialized.

However I do agree that order is important:  
Once `PYTHON3` is initialized, it defines what is returned by `vcpkg_find_acquire_program(PYTHON3)`.  
Once `MESON` is initialized, it defines what which installation of python is used to run meson.

> As a result, this python will be used in `vcpkg_install_meson`, even if neither `vcpkg_find_acquire_program(PYTHON3)` nor `vcpkg_find_acquire_program(MESON)` is used in `vcpkg_install_meson`

`vcpkg_install_meson` runs `ninja` runs build rules. The build rules define which python to use: absolute path as determined during configuration, or `python3` via environment variable `PATH` if sloppy.

> > meson running again in the build/install step (i.e. appearing in `package-...log`)? Maybe via an build/install rule...

To anser my own question (from looking at `pkgconf`): Yes, there are rules which run meson (`REGENERATE_BUILD`, but also `meson-internal__install` as a dependency of `install`). These rule use an absolute path to python.

dg0yt commented:
Regardless of the specific meson/python case, `vcpkg_find_acquire_program(script_tool)` will do an implicit `vcpkg_find_acquire_program(interpreter)`.

`vpckg-port-config.cmake` is how some behaviors are changed along port dependencies. It is a tricky game. Now that you point at `vcpkg-tool-meson`'s implementation:
 - Is it intentional that initializes `PYTHON3` for downstream ports (main effect), or does it only want to check the version (side effect)?
 - If it is intentional, wouldn't this script be the natural home for *additional* python setup? (DRY.)

(I don't know.)

vipcxj commented:
> Regardless of the specific meson/python case, vcpkg_find_acquire_program(script_tool) will do an implicit vcpkg_find_acquire_program(interpreter).

是的，我看到这一行了 [vcpkg_find_acquire_program.cmake#L81](https://github.com/microsoft/vcpkg/blob/7fd612ee1c57192d49028ed6a922e5e8bdb257f3/scripts/cmake/vcpkg_find_acquire_program.cmake#L81)

> Is it intentional that it initializes PYTHON3 for downstream ports (main effect), or does it only want to check the version (side effect)?

I think it's just trying to check the version. Expecting it to initialize downstream PYTHON3 is too subtle and exposes PYTHON3 to too big scope. Not to mention that calling `vcpkg_find_acquire_program(PYTHON3)` multiple times is no different than calling it once, and there's no need to call it again in vcpkg-configure-meson if it's intended.

vipcxj committed:
fix Build error on x64-windows #41869

Osyotr commented:
@vipcxj could you please test https://github.com/microsoft/vcpkg/pull/41874?

MonicaLiu0311 commented:
Ping @vipcxj

MonicaLiu0311 commented:
Closing this PR since it seems that no pregress is being made. Please ping us reopen if work is still being done.
